Add WebView script message and user scripts

TcT2k opened the pull request:
This follows PR #2189
It's currently implemented for Edge and WKWebView.

I'm not yet sure about the interface.

The event `wxEVT_WEBVIEW_SCRIPT_MESSAGE_RECEIVED` can be triggered via javascript like this:
* In edge: ```window.chrome.webview.postMessage('value');```
* In WKWebView and gtkwebkit ```window.webkit.messageHandlers.myobject.postMessage('value');```where `myobject` can be defined via the API and allows multiple identifiers.

The events `GetString()` contains the posted value.

Via script injection this could be brought to a unified javascript interface (as proposed in #2189). I'm not sure if that is useful or if it should at least be optional. My current interface has `wxWebView::AddScriptMessageHandler(wxString& name)` where `name` would be referenced from javascript. This method has to be called before `wxWebView::Create()` on macOS and doesn't have to be called for edge where web messages are enabled as default.

Not allowing multiple ids would sacrifice native functionality for a more common interface, but is probably worth it for a more common interface.

Native documentation:
* [Edge](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.705.50#add_webmessagereceived)
* [WKWebView](https://developer.apple.com/documentation/webkit/wkusercontentcontroller/1537172-addscriptmessagehandler?language=objc)

Any ideas on the interface are welcome.

TcT2k committed:
Add wxEVT_WEBVIEW_SCRIPT_MESSAGE_RECEIVED event

TcT2k committed:
Implement script message for edge backend

TcT2k committed:
Implement script message for WKWebView backend

vadz reviewed (commented):
Thanks for the code and the explanations!

I'm not sure neither about how best to hide the difference between Edge and WKWebView, but I think it would be better to require `AddScriptMessageHandler()` to be called under all platforms, as otherwise you would write code working just fine with Edge, but silently failing with WKWebView. It also might be worth doing something similar to what is done in [wxTextCtrlBase::OnDynamicBind()](https://github.com/wxWidgets/wxWidgets/blob/v3.1.4/src/common/textcmn.cpp#L1208) to detect this during run-time.

In normal use you'd either just call `AddScriptMessageHandler()` without parameters, which would use the default (`org.wxwidgets.something`?) name for WKWebView, or pass it an explicit name but then it would return `false` when using Edge.

I'd also add `RemoveScriptMessageHandler()` because it seems both simple to do and potentially useful.

As for unified JS interface, it seems like it would be useful, perhaps the required script could be injected the first time `AddScriptMessageHandler()` is called?

PBfordev commented on src/msw/webview_edge.cpp:
```diff
@@ -318,6 +319,20 @@ HRESULT wxWebViewEdgeImpl::OnContainsFullScreenElementChanged(ICoreWebView2* WXU
+    event.SetEventObject(m_ctrl);
+    if (SUCCEEDED(args->get_WebMessageAsJson(&msgContent)))
+        event.SetString(wxString(msgContent));
+    m_ctrl->HandleWindowEvent(event);
```
Nitpick 1:

I would be the first to admit that my coding style may be overly defensive. However, I would still always check all functions which can fail regardless of probability of that happening, and log and handle the error.

There already some "private" helper macros for [in graphicsd2d.cpp](https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/graphicsd2d.cpp#L82). TBH, I would write those macros so they do not use `wxCHECK*()` for runtime errors, I believe that should be reserved for programmer ones. I would use `wxLogApiError()` there instead, logging also the API function name that failed.

For example, something like this
```
#define wxCHECK_HRESULT_RET(hr, APIName)     \
{                                            \
   if ( FAILED(hr) )                         \
   {                                         \
         wxLogApiError(APIName, hr);         \
         return;                             \
    }                                        \
}

#define wxCHECK_HRESULT(hr, retValue, APIName) \
{                                              \
   if ( FAILED(hr) )                           \
   {                                           \
         wxLogApiError(APIName, hr);           \
         return retValue;                      \
    }                                          \
}
```
I do understand that adding such macros would be beyond the scope of this PR.

PBfordev commented:
Nitpick 2:

I wonder if `AddScriptMessageHandler()` is the best name and if it would not be more in sync with `wxWebView` API to name it `RegisterMessageHandler()`? Unfortunately, there is already is quite generic sounding `RegisterHandler()` instead of probably a bit better name like `RegisterCustomSchemeHandler()`...

TcT2k committed:
Fix unused warning

TcT2k committed:
Add script command handler to sample

TcT2k committed:
Identify script handler in message event

TcT2k committed:
Add RemoveScriptMessage handler and return bool

TcT2k committed:
Fix typo in sample

TcT2k committed:
Fix mixed EOL

TcT2k committed:
Implement wxWebView::EnableAccessToDevTools() for GTK

TcT2k committed:
Implement script message for GTK backend

TcT2k committed:
Decode web message result like RunScript() result

TcT2k committed:
Log error if script handler cannot be added

TcT2k committed:
Add named script handler to edge

TcT2k committed:
Disabled dev tools per default for Edge

This unifies the behavior among backends and is probably
the more useful default for most applications.

TcT2k committed:
Initial AddScriptMessageHandler() documentation

TcT2k committed:
Add wxWebView::AddUserScript()

This allows to inject javascript code across multiple pages

TcT2k committed:
Implement AddUserScript() for edge backend

TcT2k committed:
Implement AddUserScript() for GTK

TcT2k committed:
Add common web message identifier

TcT2k committed:
Implement AddUserScript() for WKWebView backend

TcT2k committed:
Implement common web message identifier for macOS

TcT2k committed:
Fix EOL issues

TcT2k committed:
Fix trailing whitespace

TcT2k committed:
Fix unused warning in sample

TcT2k committed:
Define unused parameters differently to improve intellisense

TcT2k committed:
Add usage documentation to wxWebView::AddScriptMessageHandler()

TcT2k committed:
Add Apple transport security to wxWebView docs

TcT2k committed:
Disable new methods for webkit1

TcT2k commented:
@PBfordev 
> I wonder if AddScriptMessageHandler() is the best name and if it would not be more in sync with wxWebView API to name it RegisterMessageHandler()? Unfortunately, there is already is quite generic sounding RegisterHandler() instead of probably a bit better named like RegisterCustomSchemeHandler()...

I've actively chosen `Add...` because I would associate a `Register..` method with something where the caller registers an object (like `RegisterHandler()`).

@vadz 
> I'm not sure neither about how best to hide the difference between Edge and WKWebView, but I think it would be better to require `AddScriptMessageHandler()` to be called under all platforms, as otherwise you would write code working just fine with Edge, but silently failing with WKWebView. It also might be worth doing something similar to what is done in [wxTextCtrlBase::OnDynamicBind()](https://github.com/wxWidgets/wxWidgets/blob/v3.1.4/src/common/textcmn.cpp#L1208) to detect this during run-time.
> 
I've changed it to require a call to `AddScriptMessageHandler()` on all platforms. The only difference is now that edge only supports a single message handler.

> In normal use you'd either just call `AddScriptMessageHandler()` without parameters, which would use the default (`org.wxwidgets.something`?) name for WKWebView, or pass it an explicit name but then it would return `false` when using Edge.

I've decided against a default name and just require the user to specify a name. This should reduce the chance for accidental naming clashes.

> 
> I'd also add `RemoveScriptMessageHandler()` because it seems both simple to do and potentially useful.
> 
Added.

> As for unified JS interface, it seems like it would be useful, perhaps the required script could be injected the first time `AddScriptMessageHandler()` is called?

The message handler is now available via `window._name_from_add_call.postMessage('messageBody')` on all platforms. This is available directly after the `AddScriptMessageHandler()` call and in all subsequent loaded pages via the newly implemented `AddUserScript()` which is certainly also useful independently of the script messages.

The sample shows the new functionality and documentation is added.
I think this should be ready to merge in it's current form (pending CI or further comments).

PBfordev commented on interface/wx/webview.h:
```diff
@@ -670,39 +731,79 @@ class wxWebView : public wxControl
+    virtual void RemoveAllUserScripts();
 
     /**
         @name Clipboard
```
Is this comment used for anything?

TcT2k replied:
These @name comments group the methods in the class documentation member list.

PBfordev commented on src/msw/webview_edge.cpp:
```diff
@@ -772,6 +837,65 @@ bool wxWebViewEdge::RunScript(const wxString& javascript, wxString* output) cons
+            wxLogApiError("RemoveScriptToExecuteOnDocumentCreated", hr);
+    }
+}
+
```
I am sorry if I am missing something but should we not clear also `m_impl->m_userScriptIds` after the loop?

BTW, AFAICT the code in this file requires C++11: would it not be better to use more readable range-based `for`s for the loops like the one above?

TcT2k replied:
You are absolutely right, changed.

PBfordev commented on interface/wx/webview.h:
```diff
@@ -433,6 +452,10 @@ class wxWebViewHandler
        and is only available in wxWidgets 3.1.5 or later.
+    @event{EVT_WEBVIEW_SCRIPT_MESSAGE_RECEIVED(id, func)}
+        Process a @c wxEVT_WEBVIEW_SCRIPT_MESSAGE_RECEIVED event
+        only available in wxWidgets 3.1.5 or later for usage details see
```
```suggestion
        only available in wxWidgets 3.1.5 or later. For usage details see
```

PBfordev commented on interface/wx/webview.h:
```diff
@@ -670,39 +731,79 @@ class wxWebView : public wxControl
+        @endcode
+
+        @param name Name of the message handler that can be used from javascript
+        @return @true if the handler could be added, @false if it could not be added
```
```suggestion
        @return @true if the handler could be added, @false if it could not be added.
```

TcT2k committed:
Apply suggestions from code review

Co-authored-by: PB <PBforDev@gmail.com>

TcT2k committed:
Clear script ids in RemoveAllUserScripts()

vadz commented:
Thanks again!

No comments from me, so I'll merge this in a few days if there are none from the others neither.

TcT2k committed:
Merge branch 'master' into webview_script_message
